system_tests/program_gas_test.go

```diff
@@ -0,0 +1,437 @@
+			// to the callee contract, which is 63/64 of the remaining gas. This happens
+			// because the tracer is evaluated before the call is executed, so the EVM
+			// doesn't know how much gas will being used.
+			//
+			// In the case of the Stylus tracer, the trace is emitted after the
+			// execution, so the EndInk field is set to the ink after the call returned.
+			// Hence, it also includes the ink spent by the callee contract.
+			//
+			// To make a precise comparison between the EVM and Stylus, we modify the
+			// EVM measurement to include the gas spent by the callee contract. To do
+			// so, we go through the opcodes after CALL until we find the first opcode
+			// in the caller's depth. Then, we subtract the gas before the call by the
+			// gas after the call returned.
+			var gasAfterCall uint64
+			for j := i + 1; j < len(result.StructLogs); j++ {
+				if result.StructLogs[j].Depth == result.StructLogs[i].Depth {
+					// back to the original call
+					gasAfterCall = result.StructLogs[j].Gas + result.StructLogs[j].GasCost
+					break
+				}
+			}
+			if gasAfterCall == 0 {
+				return nil, fmt.Errorf("malformed log: didn't get back to call original depth")
+			}
+			if i == 0 {
+				return nil, fmt.Errorf("malformed log: call is first opcode")
+			}
+			gasUsed = result.StructLogs[i-1].Gas - gasAfterCall
+		} else {
+			gasUsed = result.StructLogs[i].GasCost
+		}
+		gasUsage[op] = append(gasUsage[op], gasUsed)
+	}
+	return gasUsage, nil
+}
+
+func stylusHostiosGasUsage(ctx context.Context, rpcClient rpc.ClientInterface, tx *types.Transaction) (
+	map[string][]float64, error) {
+
+	traceOpts := struct {
+		Tracer string `json:"tracer"`
+	}{
+		Tracer: "stylusTracer",
+	}
+	var result []gethexec.HostioTraceInfo
+	err := rpcClient.CallContext(ctx, &result, "debug_traceTransaction", tx.Hash(), traceOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to trace stylus call: %w", err)
+	}
+
+	const InkPerGas = 10000
+	gasUsage := map[string][]float64{}
+	for _, hostioLog := range result {
+		gasCost := float64(hostioLog.StartInk-hostioLog.EndInk) / InkPerGas
+		gasUsage[hostioLog.Name] = append(gasUsage[hostioLog.Name], gasCost)
+	}
+	return gasUsage, nil
+}
+
+// checkPercentDiff checks whether the two values are close enough.
+func checkPercentDiff(t *testing.T, a, b float64, maxAllowedDifference float64) {
+	t.Helper()
+	if maxAllowedDifference == 0 {
+		maxAllowedDifference = 0.25
+	}
+	percentageDifference := math.Abs(a-b) / ((a + b) / 2)
+	if percentageDifference > maxAllowedDifference {
+		Fatal(t, fmt.Sprintf("gas usages are too different; got %v, max allowed is %v", percentageDifference, maxAllowedDifference))
+	}
+}
+
+// sendAndEnsureTransaction sends a transaction, ensures it succeed, and returns the total gas cost.
+func sendAndEnsureTransaction(t *testing.T, ctx context.Context, client *ethclient.Client, tx *types.Transaction) uint64 {
+	t.Helper()
```

Comment:
nit: if you hold on to builder, you can use builder.L2.SendWaitTestTransactions(t, )
I am hoping to have more tests build using these utils, and possibly adding more.. as it is I think we have too many variats of functions sending and ensuring transactions..

Reply:
just realised that SendWaitTestTransactions doesn't return receipt. Maybe we need a single tx function that does, or SendWaiTest could return an array of receipts.. feel free to either do something with it or ignore